docs/features/nullable-reference-types.md

```diff
@@ -0,0 +1,119 @@
+Nullable Reference Types
+=========
+References to reference types may be nullable, non-nullable, or null-oblivious. (The document uses the shorthand `?`, `!`, and `~`.)
+
+## Annotations
+In source, nullable reference types are annotated with `?`.
+```c#
+string? OptString; // may be null
+List<object?>? OptListOptElements; // list may be null, elements may be null
+```
+In metadata, nullable reference types are annotated with a `[Nullable]` attribute.
+```c#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class NullableAttribute : Attribute
+    {
+        public NullableAttribute() { }
+        public NullableAttribute(bool[] b) { }
+    }
+}
+```
+If annotations are enabled in a module, a `[module: Nullable]` attribute is included.
+```c#
+// C# representation of metadata
+[module: Nullable]
+[Nullable] string OptString; // string?
+[Nullable(new[] { true, true })] List<object> OptListOptElements; // List<object?>?
+```
+`NullableAttribute` cannot be used explicitly in source so the type declaration can be synthesized by the compiler if it is not included in the compilation.
+
+Unannotated reference types are non-nullable or null-oblivious depending whether the containing scope includes `[NonNullTypes]`.
+```c#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class NonNullTypesAttribute : Attribute
```

Comment:
Let's specify valid targets here too

Reply:
I don't think we can usefully apply `[NonNullTypes]` on an `enum`, but maybe we should allow it anyways, in case we add methods to enums.

---
In reply to: [194862047](https://github.com/dotnet/roslyn/pull/27721#discussion_r194862047) [](ancestors = 194862047)